feat: rest-express-auth example

gautvm opened the pull request:
This is the PR where I am working on the rest-express-auth example, details #2838. This is currently a draft.

gautvm committed:
chore: initial rest-express-auth creation

gautvm committed:
refactor: create data models

gautvm committed:
refactor: set up api structure

gautvm committed:
refactor: start some work on signup

gautvm committed:
feat(rest-express-auth) add signup route

gautvm committed:
feat(rest-express-auth) add login route

gautvm committed:
feat(rest-express-auth) added logout route

gautvm committed:
refactor(rest-express-auth) refactor isAuth middleware

gautvm committed:
feat(rest-express-auth) add account delete route

gautvm committed:
feat(rest-express-auth): posts get route

ruheni commented on typescript/rest-express-auth/package.json:
```diff
@@ -0,0 +1,23 @@
+    "express": "^4.17.1",
+    "jsonwebtoken": "^8.5.1"
+  },
+  "devDependencies": {
```
Include `prisma` dependency among the dev dependencies. It'll be used when running Prisma-specific commands when a user clones the example.

ruheni commented on typescript/rest-express-auth/.gitignore:
```diff
@@ -0,0 +1,3 @@
+node_modules/
+dist/
+*.env*
```
The `prisma-examples` repository includes a `.gitignore` file at the root of the project. This won't be necessary 🙂

ruheni commented on typescript/rest-express-auth/src/routes/Auth.ts:
```diff
@@ -0,0 +1,94 @@
+      id: req.user?.id,
+    },
+    data: {
+      sessionID: '',
```
I think we can use the `null` value here.

For more information on what setting `null` does, you can refer to this page on our docs: [Null and undefined (Reference)](https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined).

ruheni committed:
added readmes to rest-express-auth

ruheni commented:
Hi @eternalmoon1234 👋🏽

Do you have any updates on this PR? Would you like any help pushing this across the finish line? 🙂

ruheni commented:
Hey @eternalmoon1234 👋🏽

Do you intend to pick up this PR, and do you need any help?

gautvm commented:
Hey @ruheni, ✋
Yes, I do intend to pick up the PR and will finish it by the end of next week! Sorry for the inactivity with this, I've been coding on and off recently and I must say I forgot about this 😅

Looking forward to finishing this!

gautvm committed:
Merge branch 'prisma:latest' into rest-express-auth

gautvm committed:
fix: implemented fixes from review

gautvm committed:
refactor updated rest-express-auth example to follow the format of other examples

gautvm committed:
refactor: added database seeding

gautvm committed:
fix: corrected readme

gautvm commented:
Hey @ruheni 👋 

I finally finished up this PR. I've implemented the changes you proposed (as well as some others) and have followed the same standard of the [rest-express-auth](https://github.com/eternalmoon1234/prisma-examples/tree/latest/typescript/rest-express) example. 

Please review & propose any changes when you can!

gautvm committed:
fix: added isAuthenticated middleware to routes

ruheni commented on .github/readmes/typescript/rest-express-auth/README.md:
```diff
@@ -0,0 +1,40 @@
+## Contents
+
+- [Getting Started](#getting-started)
+- [Using the GraphQL API](#using-the-rest-api)
```
REST API here 😅

janpio replied:
```suggestion
- [Using the REST API](#using-the-rest-api)
```

ruheni commented on .github/readmes/typescript/rest-express-auth/README.md:
```diff
@@ -0,0 +1,40 @@
+# REST API with Authentication
+
+This example shows how to implement a **GraphQL server with TypeScript** with the following stack:
```
REST API here too.

janpio replied:
```suggestion
This example shows how to implement a **REST API with TypeScript** with the following stack:
```

ruheni commented on .github/readmes/_using-the-rest-api-auth.md:
Could you document which endpoints require authentication here? e.g documentation
- `/post`: (protected) ...

Feel free to adjust the wording as you prefer 🙂

ruheni replied:
There seems to be a content mismatch between the generated READMe and this template. Update the relevant questions to prevent the contents of the main README from being overwritten on template generation

ruheni commented on typescript/rest-express-auth/README.md:
I would recommend referring to the other templates In the `.github/readmes/typescript` folder to see what the structure is like to avoid content redundancy.

ruheni commented on typescript/rest-express-auth/prisma/seed.ts:
```diff
@@ -0,0 +1,70 @@
+const userData: Prisma.UserCreateInput[] = [
+  {
+    email: 'alice@prisma.io',
+    password: 'testPassword',
```
I would recommend the following for the test data in the event a user tries to log in with the test user:
```suggestion
    password: await argon2.hash('testPassword'),
```

ruheni reviewed (changes requested):
This looks pretty good. I've left some comments. The example seems to be missing a test for it. You can add one in the `.github/tests` folder.

One other thing I didn't include in the comments is that I think some endpoints might not require authentication, such as `/feed` and `/post/:id` (both GET endpoints). I would recommend protecting endpoints with _write_  operations in general. Let me know if you would like any help determining what endpoints to protect

gautvm committed:
fix: rest-express-auth: fixed readmes

gautvm committed:
fix: rest-express-auth: correct protected routes

gautvm committed:
refactor: rest-express-auth: add test

gautvm committed:
fix: rest-express-auth: add hashing for passwords in seeding data

gautvm commented:
Hey @ruheni 👋

I've incorporated all of your requested changes. As for the protected routes, I've modified the ones that I thought required protection based on your guidance, although please feel free to suggest any alterations if you'd propose something different.

Please let me know if there's anything else you'd like me to change 🙂

gautvm commented:
Do I need to update the `rest` test and postman collection now that there are protected routes?

ruheni commented:
Hey 👋🏾 
I would recommend creating a new test and collection for it to get the tests to pass.

gautvm commented:
Sounds good, I'll go ahead and write the new test in.

gautvm committed:
docs: rest-express-auth: add docs for isAuthenticated middleware

gautvm commented:
Hey @ruheni 👋 

The tests should pass now. Mind running the checks & verifying that everything's all good? If so, I think this is ready to merge.

XprabhatX commented:
The logs for the failed test are no longer available. Can anyone specify what's the problem.

jharrell commented:
Hello!

We are reviewing our old pull requests in this repo. If we believe that the PR is stale or abandoned, we're going to close them. 

We'd still love to have your contribution! If you'd be willing to pick this PR back up, please just comment or open a new pull request and we'll hop on it 🙏
